cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java

```diff
@@ -593,6 +604,250 @@ private Future<List<Condition>> maybeRestartConnectorTask(Reconciliation reconci
+            conditions.add(StatusUtils.buildWarningCondition("ManageOffsets", message));
+            return Future.succeededFuture(conditions);
+        }
+
+        switch (annotation) {
+            case list -> {
+                return listConnectorOffsets(reconciliation, host, apiClient, connectorName, resource, connectorSpec, conditions);
+            }
+            case alter -> {
+                return alterConnectorOffsets(reconciliation, host, apiClient, connectorName, resource, connectorSpec, conditions);
+            }
+            case reset -> {
+                return resetConnectorOffsets(reconciliation, host, apiClient, connectorName, resource, conditions);
+            }
+            default -> {
+                return Future.succeededFuture(conditions);
+            }
+        }
+    }
+
+    /**
+     * Fetches the connector offsets and writes them to a ConfigMap.
+     * This operation requires the listOffsets property to be set on
+     * the Custom Resource giving the name of the ConfigMap to use.
+     *
+     * @param reconciliation    Reconciliation marker
+     * @param host              Kafka Connect host
+     * @param apiClient         Kafka Connect REST API client
+     * @param connectorName     Name of the connector
+     * @param resource          The resource that defines the connector
+     * @param connectorSpec     Spec of the connector
+     * @param conditions        Status conditions of the resource
+     *
+     * @return Future with conditions which completes when the connector offset list action is complete
+     */
+    @SuppressWarnings({ "rawtypes" })
+    Future<List<Condition>> listConnectorOffsets(Reconciliation reconciliation, String host, KafkaConnectApi apiClient, String connectorName, CustomResource resource, KafkaConnectorSpec connectorSpec, List<Condition> conditions) {
+        Optional<ListOffsets> listOffsetsConfig = Optional.ofNullable(connectorSpec.getListOffsets());
+        if (listOffsetsConfig.isEmpty()) {
+            String message = String.format("Failed to list the connector offsets due to missing property listOffsets in %s CR.", resource.getKind());
+            LOGGER.warnCr(reconciliation, message);
+            conditions.add(StatusUtils.buildWarningCondition("ListOffsets", message));
+            return Future.succeededFuture(conditions);
+        }
+
+        String configMapName = listOffsetsConfig.get().getToConfigMap().getName();
+        return apiClient.getConnectorOffsets(reconciliation, host, port, connectorName)
+                .compose(offsets -> generateListOffsetsConfigMap(configMapName, connectorName, resource, offsets))
+                .compose(configMap -> configMapOperations.reconcile(reconciliation, resource.getMetadata().getNamespace(), configMapName, configMap))
+                .compose(v -> removeConnectorOffsetsAnnotations(reconciliation, resource))
+                .compose(v -> Future.succeededFuture(conditions), throwable -> {
```

Comment:
could be `onSuccess` and `onFailure` be used to make it more readable? No need to change just an idea.

Reply:
Using `onSuccess` and `onFailure` doesn't work because in the failure case we still want to return a succeeded future, so we need to pass a mapper, whereas `onFailure` takes a handler, so the method would still return a failed future.

I could change it to use `map` and `otherwise`?

```
        return apiClient.getConnectorOffsets(reconciliation, host, port, connectorName)
                .compose(offsets -> generateListOffsetsConfigMap(configMapName, connectorName, resource, offsets))
                .compose(configMap -> configMapOperations.reconcile(reconciliation, resource.getMetadata().getNamespace(), configMapName, configMap))
                .compose(v -> removeConnectorOffsetsAnnotations(reconciliation, resource))
                .map(v -> conditions)
                .otherwise(throwable -> {
                    // Don't fail reconciliation on error from listing offsets - add a warning and repeat list on next reconcile
                    String message = "Encountered error listing connector offsets. " + throwable.getMessage();
                    LOGGER.warnCr(reconciliation, message);
                    conditions.add(StatusUtils.buildWarningCondition("ListOffsets", message));
                    return conditions;
                });
```

Reply:
I like it, I think it's more readable but I will leave the final call to you.

Reply:
I've updated it